File: scripts/extract_vocab_to_excel.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
提取所有故事中的词汇，统计并生成Excel表格
"""

import os
import re
import json
from collections import OrderedDict
from bs4 import BeautifulSoup
import requests
import time

def extract_vocab_from_html(file_path):
    """从HTML文件中提取词汇信息"""
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    soup = BeautifulSoup(content, 'html.parser')

    # 提取词汇，格式: <span class="w">word(中文)📢</span>
    vocab_list = []
    text_divs = soup.find_all('div', class_='text')

    for text_div in text_divs:
        paragraphs = text_div.find_all('p')
        for p in paragraphs:
            # 获取完整句子
            full_text = p.get_text()

            # 查找所有词汇标记
            spans = p.find_all('span', class_='w')
            for span in spans:
                vocab_text = span.get_text()
                # 解析格式: word(中文)📢
                match = re.match(r'([a-zA-Z\-]+)\((.+?)\)📢', vocab_text)
                if match:
                    word = match.group(1).lower()
                    chinese = match.group(2)

                    # 提取例句（包含该单词的完整句子）
                    # 从完整文本中找到包含这个单词的句子
                    sentences = re.split(r'[。！？]', full_text)
                    example_sentence = ""
                    for sentence in sentences:
                        if word in sentence.lower() or vocab_text in sentence:
                            example_sentence = sentence.strip()
                            break

                    vocab_list.append({
                        'word': word,
                        'chinese': chinese,
                        'example': example_sentence
                    })

    return vocab_list

def get_phonetic_and_pos(word):
    """从有道词典API获取音标和词性"""
    url = f"https://dict.youdao.com/jsonapi?q={word}&le=en"
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }

    try:
        response = requests.get(url, headers=headers, timeout=5)
        data = response.json()

        result = {
            'phonetic': '',
            'pos': ''
        }

        # 尝试从ec词库获取
        if 'ec' in data:
            ec_data = data['ec']
            if 'word' in ec_data and len(ec_data['word']) > 0:
                word_info = ec_data['word'][0]
                # 获取音标
                if 'ukphone' in word_info:
                    result['phonetic'] = f"/{word_info.get('ukphone', '')}/"
                elif 'phone' in word_info:
                    result['phonetic'] = f"/{word_info['phone']}/"

                # 获取词性
                if 'trs' in word_info and len(word_info['trs']) > 0:
                    pos_list = []
                    for tr in word_info['trs']:
                        if 'pos' in tr and tr['pos']:
                            pos_list.append(tr['pos'])
                    if pos_list:
                        result['pos'] = ','.join(set(pos_list))

        return result
    except Exception as e:
        print(f"Error getting phonetic for {word}: {e}")
        return {'phonetic': '', 'pos': ''}

def main():
    result_dir = 'result'
    output_dir = 'statistics'

    # 确保输出目录存在
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    # 获取所有学习版HTML文件，按编号排序
    html_files = []
    for filename in os.listdir(result_dir):
        if filename.endswith('_学习版.html'):
            # 提取编号
            match = re.match(r'(\d+)_', filename)
            if match:
                number = int(match.group(1))
                html_files.append((number, filename))

    # 按编号排序
    html_files.sort(key=lambda x: x[0])

    print(f"找到 {len(html_files)} 个学习版文件")

    # 存储所有词汇
    all_vocab = OrderedDict()  # 使用OrderedDict保持插入顺序
    vocab_first_appear = {}  # 记录每个单词第一次出现的位置
    vocab_examples = {}  # 记录每个单词的例句

    # 处理每个文件
    position = 0
    for number, filename in html_files:
        file_path = os.path.join(result_dir, filename)
        print(f"处理文件: {filename}")

        vocab_list = extract_vocab_from_html(file_path)

        for vocab_info in vocab_list:
            word = vocab_info['word']
            chinese = vocab_info['chinese']
            example = vocab_info['example']

            if word not in all_vocab:
                all_vocab[word] = {
                    'word': word,
                    'chinese': chinese,
                    'count': 0,
                    'first_position': position,
                    'example': example
                }
                position += 1

            all_vocab[word]['count'] += 1

            # 如果没有例句，使用这个例句
            if word in all_vocab and not all_vocab[word]['example'] and example:
                all_vocab[word]['example'] = example

    print(f"\n共提取到 {len(all_vocab)} 个不同的词汇")

    # 获取音标和词性
    print("\n开始获取音标和词性信息...")
    total = len(all_vocab)
    for idx, (word, info) in enumerate(all_vocab.items(), 1):
        print(f"处理进度: {idx}/{total} - {word}")
        phonetic_pos = get_phonetic_and_pos(word)
        info['phonetic'] = phonetic_pos['phonetic']
        info['pos'] = phonetic_pos['pos']

        # 避免请求过于频繁
        time.sleep(0.1)

    # 生成Excel文件
    print("\n生成Excel文件...")
    try:
        import pandas as pd

        # 准备数据
        data = []
        for word, info in all_vocab.items():
            data.append({
                '单词': info['word'],
                '音标': info['phonetic'],
                '词性': info['pos'],
                '中文释义': info['chinese'],
                '出现次数': info['count'],
                '例句': info['example']
            })

        df = pd.DataFrame(data)

        # 保存为Excel
        output_file = os.path.join(output_dir, '词汇汇总表.xlsx')
        df.to_excel(output_file, index=False, engine='openpyxl')

        print(f"\n✅ 成功生成词汇汇总表: {output_file}")
        print(f"总共汇总了 {len(data)} 个词汇")

    except ImportError:
        print("\n未安装pandas，尝试使用openpyxl生成Excel...")
        try:
            from openpyxl import Workbook

            wb = Workbook()
            ws = wb.active
            ws.title = "词汇汇总"

            # 写入表头
            headers = ['单词', '音标', '词性', '中文释义', '出现次数', '例句']
            ws.append(headers)

            # 写入数据
            for word, info in all_vocab.items():
                ws.append([
                    info['word'],
                    info['phonetic'],
                    info['pos'],
                    info['chinese'],
                    info['count'],
                    info['example']
                ])

            output_file = os.path.join(output_dir, '词汇汇总表.xlsx')
            wb.save(output_file)

            print(f"\n✅ 成功生成词汇汇总表: {output_file}")
            print(f"总共汇总了 {len(all_vocab)} 个词汇")

        except ImportError:
            print("\n未安装openpyxl，生成JSON文件...")
            output_file = os.path.join(output_dir, '词汇汇总表.json')
            with open(output_file, 'w', encoding='utf-8') as f:
                json.dump(list(all_vocab.values()), f, ensure_ascii=False, indent=2)
            print(f"已生成JSON文件: {output_file}")

if __name__ == '__main__':
    main()

File: scripts/extract_vocab_to_excel.js
#!/usr/bin/env node
/**
 * 提取所有故事中的词汇，统计并生成Excel表格
 */

const fs = require('fs');
const path = require('path');
const XLSX = require('xlsx');

// 结果目录
const RESULT_DIR = path.join(__dirname, '..', 'result');
const OUTPUT_DIR = path.join(__dirname, '..', 'statistics');

// 确保输出目录存在
if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

/**
 * 从HTML文件中提取词汇信息
 */
function extractVocabFromHtml(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');

    // 提取词汇，格式: <span class="w">word(中文)📢</span>
    const vocabRegex = /<span class="w">([a-zA-Z\-]+)\((.+?)\)📢<\/span>/g;
    const vocabList = [];

    // 提取文本内容（用于获取例句）
    const textMatch = content.match(/<div class="text">(.*?)<\/div>/s);
    const textContent = textMatch ? textMatch[1] : '';

    let match;
    while ((match = vocabRegex.exec(content)) !== null) {
        const word = match[1].toLowerCase();
        const chinese = match[2];

        // 提取例句（包含该单词的句子）
        // 找到包含这个词的段落
        const paragraphRegex = new RegExp(`<p>[^<]*${match[0]}[^<]*</p>`, 'g');
        const paragraphMatch = content.match(paragraphRegex);

        let exampleSentence = '';
        if (paragraphMatch && paragraphMatch.length > 0) {
            // 提取段落中的纯文本
            const paragraphText = paragraphMatch[0]
                .replace(/<[^>]+>/g, '') // 移除HTML标签
                .replace(/[这么，那这][。！？]/g, match => match) // 分句
                .trim();

            // 找到包含单词的句子
            const sentences = paragraphText.split(/[。！？]/);
            for (const sentence of sentences) {
                if (sentence.toLowerCase().includes(word)) {
                    exampleSentence = sentence.trim();
                    break;
                }
            }
        }

        vocabList.push({
            word,
            chinese,
            example: exampleSentence
        });
    }

    return vocabList;
}

/**
 * 获取文件编号
 */
function getFileNumber(filename) {
    const match = filename.match(/^(\d+)_/);
    return match ? parseInt(match[1]) : 0;
}

/**
 * 主函数
 */
function main() {
    // 获取所有学习版HTML文件
    const files = fs.readdirSync(RESULT_DIR)
        .filter(f => f.endsWith('_学习版.html'))
        .sort((a, b) => getFileNumber(a) - getFileNumber(b));

    console.log(`找到 ${files.length} 个学习版文件`);

    // 存储所有词汇
    const allVocab = new Map();
    let position = 0;

    // 处理每个文件
    files.forEach((filename, index) => {
        const filePath = path.join(RESULT_DIR, filename);
        console.log(`处理文件 ${index + 1}/${files.length}: ${filename}`);

        const vocabList = extractVocabFromHtml(filePath);

        vocabList.forEach(vocabInfo => {
            const { word, chinese, example } = vocabInfo;

            if (!allVocab.has(word)) {
                allVocab.set(word, {
                    word,
                    chinese,
                    count: 0,
                    firstPosition: position,
                    example: example
                });
                position++;
            }

            const vocabData = allVocab.get(word);
            vocabData.count++;

            // 如果没有例句，使用这个例句
            if (!vocabData.example && example) {
                vocabData.example = example;
            }
        });
    });

    console.log(`\n共提取到 ${allVocab.size} 个不同的词汇`);

    // 准备Excel数据
    const excelData = [];
    const sortedVocab = Array.from(allVocab.values())
        .sort((a, b) => a.firstPosition - b.firstPosition);

    sortedVocab.forEach(vocab => {
        excelData.push({
            '单词': vocab.word,
            '音标': '',  // 需要单独获取
            '词性': '',  // 需要单独获取
            '中文释义': vocab.chinese,
            '出现次数': vocab.count,
            '例句': vocab.example
        });
    });

    // 生成Excel文件
    console.log('\n生成Excel文件...');
    const worksheet = XLSX.utils.json_to_sheet(excelData);

    // 设置列宽
    worksheet['!cols'] = [
        { wch: 20 }, // 单词
        { wch: 20 }, // 音标
        { wch: 15 }, // 词性
        { wch: 30 }, // 中文释义
        { wch: 10 }, // 出现次数
        { wch: 60 }  // 例句
    ];

    const workbook = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(workbook, worksheet, '词汇汇总');

    const outputFile = path.join(OUTPUT_DIR, '词汇汇总表.xlsx');
    XLSX.writeFile(workbook, outputFile);

    console.log(`\n✅ 成功生成词汇汇总表: ${outputFile}`);
    console.log(`总共汇总了 ${excelData.length} 个词汇`);
    console.log('\n注意: 音标和词性信息需要后续补充，请使用词典API或手动补充');

    // 生成JSON文件（用于后续补充音标和词性）
    const jsonFile = path.join(OUTPUT_DIR, '词汇汇总表.json');
    fs.writeFileSync(jsonFile, JSON.stringify(excelData, null, 2), 'utf-8');
    console.log(`已生成JSON文件: ${jsonFile}`);
}

main();